amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/StreamARNUtil.java

```diff
@@ -0,0 +1,90 @@
+     *                     constructed based on the provided streamName
+     * @param kinesisRegion: kinesisRegion is required to construct the stream arn
+     * @return
+     */
+    public static StreamTracker trySetStreamARN(StreamTracker streamTracker, String kinesisRegion) {
+        StsClient sts = StsClient.builder().httpClient(UrlConnectionHttpClient.builder().build()).build();
+        return trySetStreamARN(sts, streamTracker, kinesisRegion);
+    }
+
+    @VisibleForTesting
+    static StreamTracker trySetStreamARN(StsClient sts, StreamTracker streamTracker, String kinesisRegion) {
+        if (StringUtils.isEmpty(kinesisRegion)) {
+            return streamTracker;
+        }
+        Optional<Arn> stsCallerArn = getStsCallerArn(sts);
+        for (StreamConfig streamConfig : streamTracker.streamConfigList()) {
+            StreamIdentifier streamIdentifier = streamConfig.streamIdentifier();
+            if (streamIdentifier.streamARNOptional().isPresent()) {
+                continue;
+            }
+            Optional<Arn> streamArn = getStreamARNFromStsCallerArn(stsCallerArn, kinesisRegion, streamIdentifier.streamName());
+            if (!streamArn.isPresent() || !isStreamIdentifierConsistentWithARN(streamIdentifier, streamArn)) {
+                continue;
+            }
+            // Set the StreamARN
+            streamIdentifier.streamARNOptional(streamArn.map(Arn::toString));
+            log.info("StreamARN has been successfully set in the StreamTracker using STS: " + streamIdentifier.streamARNOptional());
+        }
+        return streamTracker;
+    }
+
+    private static Optional<Arn> getStsCallerArn(StsClient sts) {
+        GetCallerIdentityResponse response;
+        try {
+            response = sts.getCallerIdentity();
+        } catch (AwsServiceException | SdkClientException e) {
+            log.warn("Unable to get sts caller identity to build stream arn", e);
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(Arn.fromString(response.arn()));
+    }
+
+    private static Optional<Arn> getStreamARNFromStsCallerArn(
+            Optional<Arn> stsCallerArn, String kinesisRegion, String streamName) {
+
+        if (!stsCallerArn.isPresent() || !stsCallerArn.get().accountId().isPresent()) {
+            return Optional.empty();
+        }
+        Arn kinesisStreamArn = Arn.builder()
+                .partition(stsCallerArn.get().partition())
+                .service("kinesis")
+                .region(kinesisRegion)
+                .accountId(stsCallerArn.get().accountId().get())
+                .resource("stream/" + streamName)
+                .build();
+        return Optional.of(kinesisStreamArn);
+    }
+
+    private static boolean isStreamIdentifierConsistentWithARN(StreamIdentifier streamIdentifier, Optional<Arn> streamArn) {
+        if (streamIdentifier.accountIdOptional().isPresent() && streamArn.isPresent()) {
```

Comment:
nit: we don't need to check `streamArn.isPresent()` here. This function is called in line 44 because streamArn is present.